### PR TITLE
SI-8549 Serialization: fix regression with @SerialVersionUID / start enforcing backwards compatibility

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

**Comment by retronym** [comment]
Review by @lrytz @gkossakowski

**Comment by retronym** [comment]
I'm also considering doing a sweep though the collections to add missing `@SerialVersionUID` annotations. Of course, we first need to agree on a good criteria for "missing".

One option here is to add this to `AbstractTraversable`, as the annotation can be defined up the class hierarchy.

**Comment by retronym** [comment]
The meta discussion we can have here is how to integrate this into a Scala 2.11.1.
- Wait for our regular release date
- Release now
  - From master
  - From v2.11.0 with just this change cherry-picked

File: src/library/scala/reflect/Manifest.scala [hunk 1]
```diff
@@ -42,6 +42,9 @@ import scala.collection.mutable.{ ArrayBuilder, WrappedArray }
 @scala.annotation.implicitNotFound(msg = "No Manifest available for ${T}.")
 // TODO undeprecated until Scala reflection becomes non-experimental
 // @deprecated("Use scala.reflect.ClassTag (to capture erasures) or scala.reflect.runtime.universe.TypeTag (to capture types) or both instead", "2.10.0")
+// TODO 2.12 Use named classes, marked with `SerialVersionUID`, rather than anonymous classes for predefined class tags
+//           This will provide for better for better serialization stability. Update run/t8549.scala to check that values
+//           of types that we declare as Serializable are not instances of anoynmous classes.
```
Comment (lrytz) [inline src/library/scala/reflect/Manifest.scala]: should we log such TODOs in the issue tracker?f

**Comment by lrytz** [comment]
LGTM

File: src/library/scala/reflect/Manifest.scala (hunk 1 above)
Comment (retronym) [inline src/library/scala/reflect/Manifest.scala]: Agreed. I've opened https://issues.scala-lang.org/browse/SI-8561 and removed these comments.

**Comment by retronym** [comment]
I forgot to locally re-locker before setting the baseline for the serialization tests. That's why the tests [failed](https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/5874/console): my library was still compiled with 2.11.0, which ignored `@SerialVersionUID`. 

I just pushed a fix for the expectations.

**Comment by gkossakowski** [comment]
@retronym: I'd vote vote for 2.11.1 release done ASAP and dedicated specifically to this issue (cherry picked commit).

**Comment by lrytz** [comment]
> I'd vote vote for 2.11.1 release done ASAP and dedicated specifically to this issue 

+1

**Comment by gkossakowski** [comment]
@retronym: what's the decision on this PR?

**Comment by retronym** [comment]
We'll have another chat about it in the scala meeting tomorrow.

**Comment by adriaanm** [comment]
One concrete example of why I'm against cherry-picking: other post 2.11.0 commits are pretty relevant for 2.11.1, such as the updates to versions.properties (a9e211a284) and build.number (8301b8a84f). More generally, we should deviate from our usual flow as little as possible: simply moving up the release date has least impact. Cherry-picking may reduce the risk of regression by only including commit, but the omission of commits can also cause regressions...

**Comment by gkossakowski** [comment]
I'm ok not cherry-picking in this case.

However, I think we should work towards enabling us to cut releases around a single bug fix. We might encounter a much more serious problem in the future and we should have a process of handling such scenario. I think we are not that far from getting there.

I'll be looking into concerns like that when working on infrastructure this year.

**Comment by adriaanm** [comment]
To me, the point is that we should not merge to a release branch unless it
can go into the next release cut from that branch.
Thus, for a hot fix, submit a PR, merge it, and tag the RC, let the
automation do its thing.

On Fri, May 9, 2014 at 4:27 PM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> I'm ok not cherry-picking in this case.
> 
> However, I think we should work towards enabling us to cut releases around
> a single bug fix. We might encounter a much more serious problem in the
> future and we should have a process of handling such scenario. I think we
> are not that far from getting there.
> 
> I'll be looking into concerns like that when working on infrastructure
> this year.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3711#issuecomment-42671811
> .

## 

JOIN US. REGISTER TODAY! http://www.scaladays.org/
Scala http://www.scaladays.org/
Days http://www.scaladays.org/
June 16th-18th, http://www.scaladays.org/
Berlin http://www.scaladays.org/

**Comment by gkossakowski** [comment]
By pushing the same line of logic a little bit further, we shouldn't need RCs for minor releases. Given that we have dbuild coverage now it might be possible. I doubt we got more testing from people manually trying out RCs for minor versions.

**Comment by adriaanm** [comment]
Interesting. Let's discuss during the next meeting. We could indeed go much
faster with minor versions, simply releasing another minor instead of an RC
if a serious enough regression is found.

On Fri, May 9, 2014 at 4:34 PM, Grzegorz Kossakowski <
notifications@github.com> wrote:

> By pushing the same line of logic a little bit further, we shouldn't need
> RCs for minor releases. Given that we have dbuild coverage now it might be
> possible. I doubt we got more testing from people manually trying out RCs
> for minor versions.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3711#issuecomment-42672642
> .

## 

JOIN US. REGISTER TODAY! http://www.scaladays.org/
Scala http://www.scaladays.org/
Days http://www.scaladays.org/
June 16th-18th, http://www.scaladays.org/
Berlin http://www.scaladays.org/

**Comment by gkossakowski** [comment]
I'll remember to bring this up.

On 9 May 2014 16:37, Adriaan Moors notifications@github.com wrote:

> Interesting. Let's discuss during the next meeting. We could indeed go
> much
> faster with minor versions, simply releasing another minor instead of an
> RC
> if a serious enough regression is found.
> 
> On Fri, May 9, 2014 at 4:34 PM, Grzegorz Kossakowski <
> notifications@github.com> wrote:
> 
> > By pushing the same line of logic a little bit further, we shouldn't
> > need
> > RCs for minor releases. Given that we have dbuild coverage now it might
> > be
> > possible. I doubt we got more testing from people manually trying out
> > RCs
> > for minor versions.
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/scala/scala/pull/3711#issuecomment-42672642>
> > .
> 
> ## 
> 
> JOIN US. REGISTER TODAY! http://www.scaladays.org/
> Scala http://www.scaladays.org/
> Days http://www.scaladays.org/
> June 16th-18th, http://www.scaladays.org/
> Berlin http://www.scaladays.org/
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3711#issuecomment-42672979
> .

## 

Grzegorz Kossakowski
Scalac hacker at Typesafe http://www.typesafe.com/
twitter: @gkossakowski http://twitter.com/gkossakowski
github: @gkossakowski http://github.com/gkossakowski

**Pull request merged**